src/openrct2-ui/windows/Research.cpp

```diff
@@ -154,117 +113,218 @@ static constexpr const StringId ResearchStageNames[] = {
     STR_RESEARCH_STAGE_UNKNOWN,
 };
 
-static void WindowResearchSetPage(WindowBase* w, int32_t page);
-static void WindowResearchSetPressedTab(WindowBase* w);
-static void WindowResearchDrawTabImages(DrawPixelInfo& dpi, WindowBase* w);
-
-WindowBase* WindowResearchOpen()
+class ResearchWindow final : public Window
 {
-    WindowBase* w;
-
-    w = WindowBringToFrontByClass(WindowClass::Research);
-    if (w == nullptr)
+    void OnOpen() override
     {
-        w = WindowCreateAutoPos(WW_FUNDING, WH_FUNDING, window_research_page_events[0], WindowClass::Research, WF_10);
-        w->widgets = window_research_page_widgets[0];
-        w->number = 0;
-        w->page = 0;
-        w->frame_no = 0;
-        w->disabled_widgets = 0;
```

Comment:
Seems like this should be on a constructor of t he class `ResearchWindow()`

Reply:
The other classes I was looking at for inspiration seemed to do their initialisation through the OnOpen method not constructors. e.g. Finance and NewRide. Happy to move it, but would like to understand the reasoning, and if they should also do the same?

Reply:
That's a good question, the constructor and `OnOpen` behave very similar in a way that both are called when the window is being created. The difference is that the constructor is called first, then some of the windows parameters are set on the function `WindowCreate` and when it is almost done, it calls `OnOpen`. So I'd say that everything that can be initialised at the ctors, should. Maybe people didn't use the ctors as much because it was difficult to send down parameters? I don't know. Thoughts @IntelOrca ?

Reply:
It was difficult to send down the params so most windows aren't using the ctor approach. I'm fine either way it is done for now. We can move to using the ctor on a secondary sweep.

Reply:
All the integers have a default initialiser in the base class header, so I have left those out.